### ISSUE TITLE
docker refresh

### ISSUE THREAD

**Issue by mythi** [clearlinux/distribution#91]
This issue covers three *suggestions* that all could be handled in one go:

1. Drop 0001-Automatic-Clear-Containers-runtime-detection.patch

The OCI runtime can be selected using ```--add-runtime```. Therefore, the patch isn't absolutely necessary and makes the change to kata-runtime more difficult. Also, see 2.)

2.  Add a patch that makes OCI runtime selection an env variable

It'd much easier to change the OCI runtime setting via ```Environment=``` setting in ```docker.service```:

```
[Service]
Environment="DOCKER_RUNTIME=--add-runtime cc-runtime=/usr/bin/cc-runtime"
ExecStart=/usr/bin/dockerd $DOCKER_RUNTIME
```

This would allow the user to override just the env variable in, e.g., ```/etc/systemd/system/docker.service.d```. For example, to take the default runtime in use:

```
[Service]
Environment="DOCKER_RUNTIME=''"
```

3. Drop 0002-Use-overlay2-as-the-default-storage-driver.patch

Even without this parameter, I get:
```
# docker info | grep ^Storage
Storage Driver: overlay2
#
```

**Comment by bryteise** [comment]
I would rather keep the patch since runc containers are not preferred for the distro due to security concerns. If we force the config to virt containers via environment then the lack of fallback when the hardware doesn't support kvm is ugly.

Happy to drop the overlay2 patch though.

**Comment by mythi** [comment]
@bryteise I'm not convinced the patch is justified just to check ```/dev/kvm``` is present. IMO, the runtime selection policy belongs outside the daemon.

I tested an alternative to do this without patching (using systemd-environment-generator):
```
# cat /usr/lib/systemd/system-environment-generators/99-docker-runtime.sh 
#!/bin/bash

DOCKER_DEFAULT_RUNTIME="${DOCKER_DEFAULT_RUNTIME:-runc}"

if [ -e /dev/kvm ]; then
        DOCKER_DEFAULT_RUNTIME="cc-runtime"
fi

echo DOCKER_DEFAULT_RUNTIME=$DOCKER_DEFAULT_RUNTIME
```

with this, it's possible to:
```
# cat /usr/lib/systemd/system/docker.service.d/clearlinux.conf 
[Service]
Environment="DOCKER_EXTRA_RUNTIMES=--add-runtime cc-runtime=/usr/bin/cc-runtime"
ExecStart=
ExecStart=/usr/bin/dockerd $DOCKER_EXTRA_RUNTIMES --default-runtime=${DOCKER_DEFAULT_RUNTIME}
```

additionally, user may override ```DOCKER_DEFAULT_RUNTIME``` in, e.g., ```/etc/systemd/system/docker.service.d/custom.conf```.

**Comment by bryteise** [comment]
So I like that solution but I think we will want to measure the boot time impact of using the generator. Since these generators are a prerequisite of all units started by systemd, this solution will make everybody who has that generator on their system regardless of them having docker automatically start or not run this script.

**Comment by mythi** [comment]
True, and that's the part I don't like so much either. Another alternative I was thinking was a .service that has ```ConditionPathExists``` set to check kvm and that service just writes a config file ```docker.service``` uses in its ```EnvironmentFile```.

**Comment by bryteise** [comment]
So what if the service did the following: Check if /etc/systemd/docker.service.d/50-runtime.conf exists, if it does do nothing. Then check if /dev/kvm exists and create the 50-runtime.conf file with the environment variable setting the runtime in it.

**Comment by mythi** [comment]
untested:
```
[Unit]
Description=Set Default Docker Runtime
ConditionFirstBoot=|yes
ConditionPathExists=|/dev/kvm

[Service]
Type=oneshot
ExecStartPre=/usr/bin/mkdir -p /etc/sysconfig
ExecStart=/usr/bin/echo "DOCKER_DEFAULT_RUNTIME=\"--default-runtime cc-runtime\"" > /etc/sysconfig/docker
```

**Comment by bryteise** [comment]
So unfortunately these checks aren't enough since we might have docker without the cc-runtime. The configuration will need to move to the cc-runtime supporting package instead of docker.

**Comment by bryteise** [comment]
Okay, everything is merged at the moment, we default to the cc-runtime until after we think the kata-runtime should be the default.
